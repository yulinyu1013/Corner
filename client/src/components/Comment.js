import React, { useState } from 'react';
import { MentionsInput, Mention } from 'react-mentions';
import SingleComment from './SingleComment';
import '../styles/comment.css';
import logo from '../images/logo.png';
import { textOperation, style } from './MentionComponent';
import { createComment, updateComment } from '../fetchers/feedsApi';
import { sendNotificationsForMention } from '../fetchers/notificationApis';

function Comment({
  comment,
  pid,
  setAllComments,
  allComments,
  cornerId,
  allUsers,
  cornerName,
  isLoggedIn,
  avatar,
  userId,
  userName,
}) {
  const [commentText, setCommentText] = useState('');
  const [isediting, setIsEditing] = useState({ condition: false, id: null });
  // const userInfo = JSON.parse(sessionStorage.getItem('userInfo'));
  // const userId = userInfo ? userInfo.id : '1';
  // const userName = userInfo ? userInfo.name : 'not logged in';
  // const cornerName = 'testcornerName';
  const users = allUsers;

  const sendNotification = (nameList, postId) => {
    if (nameList.length > 0) {
      let receivers = '';
      nameList.forEach((e) => {
        receivers += `${e.id},`;
      });
      receivers = receivers.slice(0, receivers.length - 1);
      sendNotificationsForMention(userName, userId, receivers, postId, cornerId, 'comment');
    }
  };

  const changeText = (e) => {
    const innerText = e.target.value;
    setCommentText(innerText);
  };

  const submitComment = async () => {
    console.log(cornerId);
    let res = {};
    let submitText = commentText;
    if (commentText.includes('@')) {
      res = textOperation(commentText);
      submitText = res.mentionList;
      console.log(res);
    }

    if (!isediting.condition) {
      const data = await createComment(userId,
        userName,
        cornerId,
        cornerName,
        submitText,
        pid,
        avatar);
      const updatedComment = [data, ...allComments];
      setAllComments(updatedComment);
    } else {
      const updatedId = isediting.id;
      // call api to update this comment in backend
      updateComment(updatedId, submitText).then(() => {
        const updated = allComments.find((p) => p.id === updatedId);
        updated.text = submitText;
        updated.updatedAt = new Date().toISOString();
        setIsEditing({ condition: false, id: null });
      });
    }
    setCommentText('');

    if (commentText.includes('@') && res.nameList.length > 0) {
      sendNotification(res.nameList, pid);
    }
  };

  const cancelEdit = () => {
    setCommentText('');
    setIsEditing({ condition: false, id: null });
  };

  const onMentionAdd = () => {
    // on add
    // console.log(id, display);
  };

  const searchTermCallBack = (search) => {
    const arr = users.filter((u) => u.display.toLowerCase().startsWith(search.toLowerCase()));
    return arr;
  };

  const displayTransForm = (id, display) => `@${display}`;

  return (
    <div className="comment-list">
      <div className="comment-content">
        {comment.map((c) => (
          <SingleComment
            key={c.id}
            singlecomment={c}
            setIsEditing={setIsEditing}
            setCommentText={setCommentText}
            allComments={allComments}
            setAllComments={setAllComments}
            userId={userId}
          />
        ))}
      </div>
      {isLoggedIn && (
        <div className="comment-box">
          <img className="user" src={avatar || logo} alt="user" />
          <MentionsInput
            className="input"
            value={commentText}
            placeholder="Write a comment..."
            onChange={changeText}
            singleLine
            style={style}
            allowSuggestionsAboveCursor
          >
            <Mention
              trigger="@"
              data={(search) => searchTermCallBack(search)}
              onAdd={onMentionAdd}
              appendSpaceOnAdd
              markup="@[__display__]*[__id__]"
              displayTransform={(id, display) => displayTransForm(id, display)}
            />
          </MentionsInput>
          <button type="submit" onClick={submitComment}>{isediting.condition ? 'UPDATE' : 'POST'}</button>
          {isediting.condition && <button className="cancel-edit-button" type="submit" onClick={cancelEdit}>|CANCEL</button>}
        </div>
      )}
    </div>
  );
}

export default Comment;
