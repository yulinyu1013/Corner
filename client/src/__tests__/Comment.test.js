import React from 'react';
import renderer from 'react-test-renderer';
import Comment from '../components/Comment';
import user from '../images/user.png';

const comment = [
  {
    id: 3,
    creator: 1,
    creatorName: 'Jhon',
    corner: 'test_corner',
    cornerName: 'test_corner',
    type: 'comment',
    pid: 1,
    text: 'first test comment',
    pic: [],
    audio: null,
    video: null,
    createdAt: '2021-08-16T23:00:33.010+02:00',
    updatedAt: '2021-08-16T23:00:33.010+02:00',
    isFlagged: 0,
    isHidden: 0,
  },
  {
    id: 4,
    creator: 1,
    creatorName: 'Jhon',
    corner: 'test_corner',
    cornerName: 'test_corner',
    type: 'comment',
    pid: 1,
    text: 'second test comment',
    pic: [],
    audio: null,
    video: null,
    createdAt: '2021-08-16T23:00:33.010+02:00',
    updatedAt: '2021-08-16T23:00:33.010+02:00',
    isFlagged: 0,
    isHidden: 0,
  },
];

const noComment = [];

it('renders Comment correctly when user loggedin and is admin', () => {
  const tree = renderer.create(<Comment comment={comment} pid='1' isLoggedIn avatar={user} cornerid='1' userName='test' userId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Comment correctly when no comment', () => {
  const tree = renderer.create(<Comment comment={noComment} pid='1' isLoggedIn avatar={user} cornerid='1' userName='test' userId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Comment correctly when no avatar', () => {
  const tree = renderer.create(<Comment comment={comment} pid='1' isLoggedIn cornerid='1' userName='test' userId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Comment correctly when user not loggedin', () => {
  const tree = renderer.create(<Comment comment={comment} pid='1' isLoggedIn={false} cornerid='1' userName='test' userId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Comment correctly when user is not creator', () => {
  const tree = renderer.create(<Comment comment={comment} pid='1' isLoggedIn={false} cornerid='1' userName='test' userId='5' />).toJSON();
  expect(tree).toMatchSnapshot();
});