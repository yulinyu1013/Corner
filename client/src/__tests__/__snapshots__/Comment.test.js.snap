// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Comment correctly when no avatar 1`] = `
<div
  className="comment-list"
>
  <div
    className="comment-content"
  >
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            first test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            second test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
  </div>
  <div
    className="comment-box"
  >
    <img
      alt="user"
      className="user"
      src="logo.png"
    />
    <div
      className="input input--singleLine"
      style={
        Object {
          "overflowY": "visible",
          "position": "relative",
        }
      }
    >
      <div
        className="input__control"
        style={
          Object {
            "border": "none",
            "display": "inline-block",
            "fontSize": 10,
          }
        }
      >
        <div
          className="input__highlighter input__highlighter--singleLine"
          style={
            Object {
              "border": "1px solid transparent",
              "boxSizing": "border-box",
              "color": "transparent",
              "overflow": "hidden",
              "position": "relative",
              "textAlign": "start",
              "whiteSpace": "pre",
              "width": "100%",
              "wordWrap": null,
            }
          }
        >
           
        </div>
        <input
          className="input__input"
          onBlur={[Function]}
          onChange={[Function]}
          onCompositionEnd={[Function]}
          onCompositionStart={[Function]}
          onKeyDown={[Function]}
          onScroll={[Function]}
          onSelect={[Function]}
          placeholder="Write a comment..."
          style={
            Object {
              "backgroundColor": "transparent",
              "border": "1px solid #dfdfdf",
              "borderRadius": 20,
              "boxSizing": "border-box",
              "display": "block",
              "fontFamily": "inherit",
              "fontSize": "inherit",
              "left": 0,
              "letterSpacing": "inherit",
              "margin": 0,
              "outline": "none",
              "padding": 10,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      onClick={[Function]}
      type="submit"
    >
      POST
    </button>
  </div>
</div>
`;

exports[`renders Comment correctly when no comment 1`] = `
<div
  className="comment-list"
>
  <div
    className="comment-content"
  />
  <div
    className="comment-box"
  >
    <img
      alt="user"
      className="user"
      src="user.png"
    />
    <div
      className="input input--singleLine"
      style={
        Object {
          "overflowY": "visible",
          "position": "relative",
        }
      }
    >
      <div
        className="input__control"
        style={
          Object {
            "border": "none",
            "display": "inline-block",
            "fontSize": 10,
          }
        }
      >
        <div
          className="input__highlighter input__highlighter--singleLine"
          style={
            Object {
              "border": "1px solid transparent",
              "boxSizing": "border-box",
              "color": "transparent",
              "overflow": "hidden",
              "position": "relative",
              "textAlign": "start",
              "whiteSpace": "pre",
              "width": "100%",
              "wordWrap": null,
            }
          }
        >
           
        </div>
        <input
          className="input__input"
          onBlur={[Function]}
          onChange={[Function]}
          onCompositionEnd={[Function]}
          onCompositionStart={[Function]}
          onKeyDown={[Function]}
          onScroll={[Function]}
          onSelect={[Function]}
          placeholder="Write a comment..."
          style={
            Object {
              "backgroundColor": "transparent",
              "border": "1px solid #dfdfdf",
              "borderRadius": 20,
              "boxSizing": "border-box",
              "display": "block",
              "fontFamily": "inherit",
              "fontSize": "inherit",
              "left": 0,
              "letterSpacing": "inherit",
              "margin": 0,
              "outline": "none",
              "padding": 10,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      onClick={[Function]}
      type="submit"
    >
      POST
    </button>
  </div>
</div>
`;

exports[`renders Comment correctly when user is not creator 1`] = `
<div
  className="comment-list"
>
  <div
    className="comment-content"
  >
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            first test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            second test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders Comment correctly when user loggedin and is admin 1`] = `
<div
  className="comment-list"
>
  <div
    className="comment-content"
  >
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            first test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            second test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
  </div>
  <div
    className="comment-box"
  >
    <img
      alt="user"
      className="user"
      src="user.png"
    />
    <div
      className="input input--singleLine"
      style={
        Object {
          "overflowY": "visible",
          "position": "relative",
        }
      }
    >
      <div
        className="input__control"
        style={
          Object {
            "border": "none",
            "display": "inline-block",
            "fontSize": 10,
          }
        }
      >
        <div
          className="input__highlighter input__highlighter--singleLine"
          style={
            Object {
              "border": "1px solid transparent",
              "boxSizing": "border-box",
              "color": "transparent",
              "overflow": "hidden",
              "position": "relative",
              "textAlign": "start",
              "whiteSpace": "pre",
              "width": "100%",
              "wordWrap": null,
            }
          }
        >
           
        </div>
        <input
          className="input__input"
          onBlur={[Function]}
          onChange={[Function]}
          onCompositionEnd={[Function]}
          onCompositionStart={[Function]}
          onKeyDown={[Function]}
          onScroll={[Function]}
          onSelect={[Function]}
          placeholder="Write a comment..."
          style={
            Object {
              "backgroundColor": "transparent",
              "border": "1px solid #dfdfdf",
              "borderRadius": 20,
              "boxSizing": "border-box",
              "display": "block",
              "fontFamily": "inherit",
              "fontSize": "inherit",
              "left": 0,
              "letterSpacing": "inherit",
              "margin": 0,
              "outline": "none",
              "padding": 10,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      onClick={[Function]}
      type="submit"
    >
      POST
    </button>
  </div>
</div>
`;

exports[`renders Comment correctly when user not loggedin 1`] = `
<div
  className="comment-list"
>
  <div
    className="comment-content"
  >
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            first test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
    <div
      className="single-comment"
    >
      <div
        className="top-content"
      >
        <img
          alt="commentUser"
          className="comment-pic"
          src="logo.png"
        />
        <div
          className="grey-box"
        >
          <p>
            Jhon
          </p>
          <p>
            second test comment
          </p>
        </div>
      </div>
      <div
        className="comment-operation"
      >
        <p>
          8/16/2021
        </p>
      </div>
    </div>
  </div>
</div>
`;
